gpytorch/variational/variational_strategy.py

```diff
@@ -116,16 +119,20 @@ def forward(self, x):
 
             # Compute predictive covariance
             predictive_covar = data_data_covar
-            if beta_features.fast_pred_var.on():
+            if not self.training and beta_features.fast_pred_var.on():
                 correction = RootLazyTensor(data_induc_covar.matmul(self.prior_root_inv)).mul(-1)
                 correction = correction + RootLazyTensor(data_induc_covar.matmul(self.variational_root))
                 predictive_covar = predictive_covar + correction
             else:
                 induc_data_covar = induc_data_covar.evaluate()
                 inv_product = induc_induc_covar.inv_matmul(induc_data_covar)
                 factor = variational_dist.lazy_covariance_matrix.root_decomposition().matmul(inv_product)
-                right_factor = factor - inv_product
-                left_factor = (factor - induc_data_covar).transpose(-1, -2)
-                predictive_covar = predictive_covar + MatmulLazyTensor(left_factor, right_factor)
+                predictive_covar = RootLazyTensor(factor.transpose(-2, -1))
+
+                if gpytorch.beta_features.diagonal_correction.on():
+                    fake_diagonal = (inv_product * induc_data_covar).sum(0)
+                    real_diagonal = data_data_covar.diag()
+                    diag_correction = DiagLazyTensor(real_diagonal - fake_diagonal)
```

Comment:
Be sure to clamp `real_diagonal - fake_diagonal` - otherwise you could get NaN errors.